### ISSUE TITLE
puzzled about "discriminator.trainable=False" 

### ISSUE THREAD

**Issue by wangqin4377** [eriklindernoren/Keras-GAN#73]
in pix2pix.py:

`self.discriminator = self.build_discriminator()`
`self.discriminator.compile(loss='mse', optimizer=optimizer, metrics=['accuracy'])`

`# For the combined model we will only train the generator`
`self.discriminator.trainable = False`
`# Discriminators determines validity of translated images / condition pairs`
`valid = self.discriminator([fake_A, img_B])`
`self.combined = Model(inputs=[img_A, img_B], outputs=[valid, fake_A]) `

The D network part of the combined model has same instance with real D network(not deep copy), and set  `discriminator.trainable = False` will disable both combined model and real D model trainable. So shall there has a switch to switch enable/disable the discriminator model when it's on training ?

**Comment by ahrnbom** [comment]
I'm a bit puzzled by this as well. I *think* that it works like this: When you call `compile`, it builds a trainable model, and uses the current `trainable` flags. A compiled model can then not have its `trainable` flags changed, so we are free to change them and compile another model with different flags.

Then, for some reason, those two compiled models can still have the same weights (I guess?) even though tensorflow/keras still sees them as clearly separated things.

Somebody who knows more about this, please chime in and correct me if I'm wrong about this.

**Comment by eriklindernoren** [comment]
@ahrnbom You're exactly right . By setting `trainable=False` after the discriminator has been compiled the discriminator is still trained during `discriminator.train_on_batch` but since it's set to non-trainable before the combined model is compiled it's not trained during `combined.train_on_batch`.

**Comment by ahrnbom** [comment]
On my setup I get this warning from Keras:

```/opt/conda/lib/python3.5/site-packages/keras/engine/training.py:940: UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
  'Discrepancy between trainable weights and collected trainable'
```

To me at least this makes me think there's probably a nicer way to achieve this.

**Comment by eriklindernoren** [comment]
I hear you. However, the same is done in [keras/examples](https://github.com/keras-team/keras/blob/master/examples/mnist_acgan.py#L138). 

```python
    # build the discriminator
    print('Discriminator model:')
    discriminator = build_discriminator()
    discriminator.compile(
        optimizer=Adam(lr=adam_lr, beta_1=adam_beta_1),
        loss=['binary_crossentropy', 'sparse_categorical_crossentropy']
    )
    discriminator.summary()

    # build the generator
    generator = build_generator(latent_size)

    latent = Input(shape=(latent_size, ))
    image_class = Input(shape=(1,), dtype='int32')

    # get a fake image
    fake = generator([latent, image_class])

    # we only want to be able to train generation for the combined model
    discriminator.trainable = False
    fake, aux = discriminator(fake)
    combined = Model([latent, image_class], [fake, aux])
```

**Comment by wangqin4377** [comment]
The first question I got it, thanks. And the warning exactly proves that set `trainable=False` not affect the original discriminator model.

**Comment by kjaquier** [comment]
See https://github.com/keras-team/keras/issues/8585#issuecomment-428815858

**Comment by thusinh1969** [comment]
> @ahrnbom You're exactly right . By setting `trainable=False` after the discriminator has been compiled the discriminator is still trained during `discriminator.train_on_batch` but since it's set to non-trainable before the combined model is compiled it's not trained during `combined.train_on_batch`.

I have scratched my head for 1 full day on this. Thanks 👍
